### PR TITLE
Add multiple dofmaps and elements to Geometry

### PR DESCRIPTION
**Pull request by chrisrichardson**
Updates dofmapbuilder to build multiple dofmaps (on the same range) using multiple element dof layouts, when there is a mixed topology.

### CODE REVIEW

File: python/test/unit/mesh/test_mixed_topology.py [hunk 1]
```diff
@@ -169,6 +189,37 @@ def test_parallel_mixed_mesh():
     assert topology.index_maps(2)[0].size_global == size * 2
     assert topology.index_maps(2)[1].size_global == size
 
+    # Create dofmaps for Geometry
+    tri = coordinate_element(CellType.triangle, 1)
+    quad = coordinate_element(CellType.quadrilateral, 1)
+    nodes = [3 * rank + i for i in range(6)]
+    xdofs = np.array([0, 1, 4, 0, 3, 4, 1, 4, 2, 5], dtype=int) + 3 * rank
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mixed_topology.py:196]: `int32` or `int64`?

Comment (chrisrichardson) [reply python/test/unit/mesh/test_mixed_topology.py:196]: Yes

File: cpp/dolfinx/fem/utils.cpp [hunk 2]
```diff
@@ -133,6 +133,55 @@ fem::DofMap fem::create_dofmap(
   return DofMap(layout, index_map, bs, std::move(dofmaps.front()), bs);
 }
 //-----------------------------------------------------------------------------
+std::vector<fem::DofMap> fem::create_dofmaps(
+    MPI_Comm comm, const std::vector<ElementDofLayout>& layouts,
+    mesh::Topology& topology,
+    std::function<void(const std::span<std::int32_t>&, std::uint32_t)>
```
Comment (garth-wells) [inline cpp/dolfinx/fem/utils.cpp:139]: I've removed all reference versions of `std::span` elsewhere.

File: cpp/dolfinx/fem/utils.cpp [hunk 3]
```diff
@@ -133,6 +133,55 @@ fem::DofMap fem::create_dofmap(
...
+  // DOF numbering on each cell
+  if (unpermute_dofs)
+  {
+    if (layouts.size() != 1)
+    {
+      throw std::runtime_error(
+          "DOF transformations not yet supported in mixed topology.");
+    }
+    const int D = topology.dim();
+    const int num_cells = topology.connectivity(D, 0)->num_nodes();
```
Comment (garth-wells) [inline cpp/dolfinx/fem/utils.cpp:166]: `int` type?

Let's be very careful with `int` types. We've taken a lot of care to get it right.

File: cpp/dolfinx/fem/utils.cpp [hunk 4]
```diff
@@ -133,6 +133,55 @@ fem::DofMap fem::create_dofmap(
...
+        = topology.get_cell_permutation_info();
+    int dim = layouts.front().num_dofs();
+    for (std::int32_t cell = 0; cell < num_cells; ++cell)
+    {
+      std::span<std::int32_t> dofs(dofmaps.front().data() + cell * dim, dim);
+      unpermute_dofs(dofs, cell_info[cell]);
+    }
+  }
+
+  std::vector<DofMap> dms;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/utils.cpp:178]: Use a better name, e.g. `dofmaps`.

File: cpp/dolfinx/fem/utils.h [hunk 5]
```diff
@@ -246,6 +246,22 @@ DofMap create_dofmap(
     std::function<std::vector<int>(const graph::AdjacencyList<std::int32_t>&)>
         reorder_fn);
 
+/// @brief Create a set of dofmaps on mesh
```
Comment (garth-wells) [inline cpp/dolfinx/fem/utils.h:249]: Where are the meshes? On different topologies?

File: cpp/dolfinx/fem/utils.h [hunk 6]
```diff
@@ -246,6 +246,22 @@ DofMap create_dofmap(
...
+/// @param[in] layouts Dof layout on each element type
+/// @param[in] topology Mesh topology
+/// @param[in] unpermute_dofs Function to un-permute dofs. `nullptr`
+/// when transformation is not required.
+/// @param[in] reorder_fn Graph reordering function called on the dofmaps
+/// @return The list of new dof maps
+std::vector<DofMap> create_dofmaps(
+    MPI_Comm comm, const std::vector<ElementDofLayout>& layouts,
+    mesh::Topology& topology,
+    std::function<void(const std::span<std::int32_t>&, std::uint32_t)>
```
Comment (garth-wells) [inline cpp/dolfinx/fem/utils.h:260]: Remove references for spans.

File: python/dolfinx/wrappers/mesh.cpp [hunk 7]
```diff
@@ -534,6 +545,18 @@ void mesh(nb::module_& m)
           { return MPICommWrapper(self.comm()); },
           nb::keep_alive<0, 1>());
 
+  m.def("create_geometry",
+        [](const dolfinx::mesh::Topology& topology,
+           const std::vector<dolfinx::fem::CoordinateElement<double>>& elements,
+           const std::vector<std::int64_t>& nodes,
```
Comment (garth-wells) [inline python/dolfinx/wrappers/mesh.cpp:551]: Use `ndarray` for `nodes`, `xdofs` and `x`.

File: python/test/unit/fem/test_mixed_mesh_dofmap.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        dofmaps += [ufcx_dofmap]
+
+    cpp_dofmaps = _cpp.fem.create_dofmaps(
+        mesh.comm, [ffi.cast("uintptr_t", ffi.addressof(dm)) for dm in dofmaps], mesh.topology
+    )
+
+    return (cpp_elements, cpp_dofmaps)
+
+
+def test_el_dm():
```
Comment (garth-wells) [inline python/test/unit/fem/test_mixed_mesh_dofmap.py:37]: No idea what this function name means.

File: python/test/unit/fem/test_mixed_mesh_dofmap.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    set_log_level(LogLevel.INFO)
+    geom = create_geometry(
+        topology, [tri._cpp_object, quad._cpp_object], nodes, xdofs, x.flatten(), 2
+    )
+    mesh = Mesh_float64(MPI.COMM_WORLD, topology, geom)
+
+    assert mesh.geometry.x.shape == (6, 3)
+
+    # Second order dofmap on mixed mesh
+    el, dm = create_element_dofmap(mesh, [basix.CellType.triangle, basix.CellType.quadrilateral], 2)
```
Comment (garth-wells) [inline python/test/unit/fem/test_mixed_mesh_dofmap.py:78]: `el` and `dm`?

File: python/test/unit/fem/test_mixed_mesh_dofmap.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    print(q0.list)
+    assert q0.list.shape == (2, 6)
+    assert len(q0.dof_layout.entity_dofs(2, 0)) == 0
+    # Quadrilaterals
+    print(q1.list)
+    assert q1.list.shape == (1, 9)
+    assert len(q1.dof_layout.entity_dofs(2, 0)) == 1
+
+
+def test_el_dm_prism():
```
Comment (garth-wells) [inline python/test/unit/fem/test_mixed_mesh_dofmap.py:98]: ???

File: python/test/unit/mesh/test_mixed_topology.py [hunk 11]
```diff
@@ -1,10 +1,17 @@
...
+import numpy as np
+
+from dolfinx.cpp.log import set_thread_name
+from dolfinx.cpp.mesh import Mesh_float64, create_geometry, create_topology
+from dolfinx.fem import coordinate_element
+from dolfinx.log import LogLevel, set_log_level
 from dolfinx.mesh import CellType, GhostMode, create_unit_cube
 
 
 def test_triquad():
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mixed_topology.py:12]: ???

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/utils.cpp (hunk 4 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/utils.cpp:178]: `dofmaps` already used in this function - and this is very local scope, just a few lines. I think it is pretty clear.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 12]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
 /// Build a simple dofmap from ElementDofmap based on mesh entity
 /// indices (local and global)
 ///
 /// @param [in] mesh The mesh to build the dofmap on
 /// @param [in] topology The mesh topology
 /// @param [in] element_dof_layout The layout of dofs on each cell type
-/// @return Returns: * dofmap for first element type [0] (local to the process)
+/// @return Returns: * dofmaps for each cell type (local to the process) as
+/// (width, [dofs])
 ///                  * local-to-global map for each local dof
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:147]: The last two bullet points are ambiguous. Need to be very precise of 'local' and 'global' - overloaded phrases in the context of dofmaps.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 13]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
 ///                  * local-to-entity map for each local dof
-/// Entities are represented as {dimension, mesh entity index}.
-std::tuple<dofmap_t, std::vector<std::int64_t>,
-           std::vector<std::pair<std::int8_t, std::int32_t>>>
-build_basic_dofmap(const mesh::Topology& topology,
-                   const fem::ElementDofLayout& element_dof_layout)
+///                  * index maps for each entity type in dofmaps
+///                  * global offset for dofs on this process
+/// Entities are represented as {index_map number, mesh entity index}.
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:151]: What does this line relate to?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 14]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
   // Topological dimension
   const std::size_t D = topology.dim();
+  const std::size_t num_cell_types = topology.entity_types(D).size();
+
+  LOG(INFO) << "Checking required entities per dimension";
+
+  // Find which dimensions (d) and entity types (et) are required
+  // and the number of dofs which are required for each (d, et) combination.
+  // Also store the IndexMaps and local offsets for each.
+  std::vector<std::pair<std::int8_t, std::int8_t>> required_dim_et;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:172]: Way a `std::pair` when types are the same (`std::int8_t`)? 

`et` isn't informative.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 15]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
-  // the number mesh mesh entities (that are required), and
-  // and compute number of dofs on this process
-  std::vector<std::int8_t> needs_entities(D + 1, false);
-  std::vector<std::int32_t> num_entities(D + 1, 0);
-  std::int32_t local_size = 0;
-  for (std::size_t d = 0; d <= D; ++d)
+  for (std::size_t i = 0; i < num_cell_types; ++i)
   {
-    if (element_dof_layout.num_entity_dofs(d) > 0)
+    mesh::CellType cell_type = topology.entity_types(D)[i];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:179]: It's a bit ugly that `topology.entity_types(D)` creates a new `std::vector` each time.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 16]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+    auto entity_dofs = element_dof_layouts[i].entity_dofs_all();
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:180]: Avoid `auto` here - with `auto` it's not clear what the type is.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 17]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+
+    for (std::size_t d = 0; d <= D; ++d)
     {
-      if (!topology.connectivity(d, 0))
+      const std::vector<std::vector<int>>& edd = entity_dofs[d];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:184]: `edd`?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 18]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+      auto entity_types_d = topology.entity_types(d);
+      for (std::size_t e = 0; e < edd.size(); ++e)
       {
-        throw std::runtime_error(
-            "Cannot create basic dofmap. Missing entities of dimension "
-            + std::to_string(d) + " .");
+        if (edd[e].size() > 0)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:188]:
```suggestion
        if (!edd[e].empty())
```

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 19]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+        {
+          // There is a dof on this entity... find entity type index
+          mesh::CellType et = mesh::cell_entity_type(cell_type, d, e);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:191]: `et`?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 20]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+            const std::int32_t num_entity_dofs = edd[e].size();
+            num_entity_dofs_et.push_back(num_entity_dofs);
+            auto im = topology.index_maps(d)[et_index];
+            topo_index_maps.push_back(im);
+            local_entity_offsets.push_back(
+                local_entity_offsets.back()
+                + num_entity_dofs * (im->size_local() + im->num_ghosts()));
+
+            if (d < D and !topology.connectivity({D, i}, {d, et_index}))
+              throw std::runtime_error("Missing needed connectivity. Cell type:"
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:213]: Add curly braces.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 21]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+                                       + std::to_string(i)
+                                       + "to dim:" + std::to_string(d)
+                                       + ", ent:" + std::to_string(et_index));
+          }
+          else
+          {
+            int k = std::distance(required_dim_et.begin(), find_dim_et);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:220]:
```suggestion
            std::size_t k = std::distance(required_dim_et.begin(), find_dim_et);
```

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 22]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+    LOG(INFO) << s.str();
+  }
+#endif
 
-    // Iterate over each topological dimension for this element
-    std::int32_t offset_local = 0;
-    for (std::size_t d = 0; d <= D; ++d)
+  // Dofmaps on each cell type as (width, [cell_dofs])
+  std::vector<dofmap_t> dofs(num_cell_types);
+
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:244]:
```suggestion
```

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 23]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+  for (std::size_t i = 0; i < num_cell_types; ++i)
+  {
+    mesh::CellType cell_type = topology.entity_types(D)[i];
+    const std::vector<std::vector<std::vector<int>>> entity_dofs
+        = element_dof_layouts[i].entity_dofs_all();
+    assert(entity_dofs.size() == D + 1);
+
+    // Loop over cells of this type, and build dofmap
+    std::int32_t num_cells = topology.index_maps(D)[i]->size_local()
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:253]: This creates `topology.index_maps(D)[i]` twice.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 24]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
-      if (needs_entities[d])
+      // Wrap dofs for cell c
+      std::span<std::int32_t> dofs_c(dofs[i].array.data() + dofmap_offset,
+                                     dofmap_width);
+      dofmap_offset += dofmap_width;
+
+      // Iterate over required entities for this element, dimension and type
+      for (std::size_t k = 0; k < required_dim_et.size(); ++k)
       {
+        std::size_t d = required_dim_et[k].first;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:272]: `d`?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 25]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+        std::size_t et = required_dim_et[k].second;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:273]: `et`?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 26]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+        mesh::CellType e_type = topology.entity_types(d)[et];
+
         const std::vector<std::vector<int>>& e_dofs_d = entity_dofs[d];
 
-        // Iterate over each entity of current dimension d
-        std::size_t num_entity_dofs = e_dofs_d[0].size();
+        // Iterate over each entity of current dimension d and type et
         const std::int32_t* c_to_e
-            = (d == D) ? nullptr : connectivity[d]->links(c).data();
+            = ((std::size_t)d == D)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:280]: This is hard to read. Why not use a `std::span`?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 27]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+          // Use separate connectivity index 'w' which only advances for
+          // correct entities
+          if (mesh::cell_entity_type(cell_type, d, e) == e_type)
           {
-            int dof_local = e_dofs_d[e][i];
-            dofs_c[dof_local]
-                = offset_local + num_entity_dofs * e_index_local + i;
+            std::size_t num_entity_dofs = e_dofs_d[e].size();
+            assert((int)num_entity_dofs == num_entity_dofs_et[k]);
+            std::int32_t e_index_local = ((std::size_t)d == D) ? c : c_to_e[w];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:294]:
```suggestion
            std::int32_t e_index_local = (std::size_t)d == D ? c : c_to_e[w];
```
Get rid of the cast?

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 28]
```diff
@@ -136,147 +136,223 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
+            ++w;
+
+            // Loop over dofs belonging to entity e of dimension d (d, e)
+            // d: topological dimension
+            // e: local entity index
+            // dof_local: local index of dof at (d, e)
+            for (std::size_t j = 0; j < num_entity_dofs; ++j)
+            {
+              int dof_local = e_dofs_d[e][j];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:303]: Is this the right integer type?

Avoid double-indexing inside the `j` loop.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 29]
```diff
@@ -306,14 +382,10 @@ std::pair<std::vector<std::int32_t>, std::int32_t> compute_reordering_map(
...
+  // Get mesh entity ownership offset for each IndexMap
   std::vector<std::int32_t> offset(index_maps.size(), -1);
-  for (std::size_t d = 0; d < offset.size(); ++d)
-  {
-    auto map = index_maps[d];
-    if (map)
-      offset[d] = map->size_local();
-  }
+  std::transform(index_maps.begin(), index_maps.end(), offset.begin(),
+                 [](const auto& im) { return im->size_local(); });
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:388]:
```suggestion
                 [](auto& map) { return map->size_local(); });
```

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 30]
```diff
@@ -550,56 +619,44 @@ fem::build_dofmap_data(
...
-  for (std::size_t i = 0; i < dofmaps.front().size(); ++i)
-    dofmaps.front()[i] = old_to_new[node_graph0.array[i]];
+  // Build re-ordered dofmaps
+  std::vector<std::vector<std::int32_t>> dofmaps(node_graphs.size());
+  for (std::size_t i = 0; i < dofmaps.size(); ++i)
+  {
+    dofmaps[i].resize(node_graphs[i].array.size());
+    for (std::size_t j = 0; j < node_graphs[i].array.size(); ++j)
+    {
+      std::int32_t old_node = node_graphs[i].array[j];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:656]: Avoid double indexing.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Part way through . . .

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 14 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:172]: I guess I could use `std::array<std::int8_t, 2>` but I think `pair` shows the intent better, as the two values refer to different things.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 15 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:179]: I'll move outside the loop. It is only a small vector, and it works better with Python like this.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 16 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:180]: `const std::vector<std::vector<std::vector<int>>>&` - it's just a bit of a mouthful.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 17 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:184]: OK, I will rename this one.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 24 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:271]: Used throughout for topological dimension. I will add a comment here.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 28 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:302]: `int` is used throughout `ElementDofLayout` - we can change it in another PR.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 31]
```diff
@@ -136,147 +136,227 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
-            + std::to_string(d) + " .");
+        if (!entity_dofs_d[e].empty())
+        {
+          // There is a dof on this entity... find entity type index
+          auto et_it = std::find(entity_types[d].begin(), entity_types[d].end(),
+                                 mesh::cell_entity_type(cell_type, d, e));
+          assert(et_it != entity_types[d].end());
+          int et_index = std::distance(entity_types[d].begin(), et_it);
+
+          auto find_dim_et
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:199]: `find_dim_et` is a bit cryptic.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 32]
```diff
@@ -136,147 +136,227 @@ reorder_owned(const std::vector<dofmap_t>& dofmaps, std::int32_t owned_size,
...
-  for (std::int32_t c = 0; c < num_cells; ++c)
+#ifndef NDEBUG
   {
-    // Wrap dofs for cell c
-    std::span dofs_c(dofs.array.data() + c * dofs.width, dofs.width);
+    // Debug output
+    std::stringstream s;
+    s << "Required entities:";
+    for (std::size_t i = 0; i < required_dim_et.size(); ++i)
+      s << "(" << (int)required_dim_et[i].first << ", "
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:239]: Add curly braces.

File: cpp/dolfinx/fem/dofmapbuilder.cpp [hunk 33]
```diff
@@ -438,39 +513,37 @@ std::pair<std::vector<std::int64_t>, std::vector<int>> get_global_indices(
...
+    MPI_Neighbor_allgather(&num_indices, 1, MPI_INT, size_recv.data(), 1,
+                           MPI_INT, comm[d]);
+
+    // Compute displacements for data to receive. Last entry has total
+    // number of received items.
+    disp_recv[d].resize(src.size() + 1);
+    std::partial_sum(size_recv.begin(), size_recv.begin() + src.size(),
+                     disp_recv[d].begin() + 1);
+
+    // TODO: use MPI_Ineighbor_alltoallv
```
Comment (garth-wells) [inline cpp/dolfinx/fem/dofmapbuilder.cpp:539]: Can this be done now?

**Review by garth-wells — COMMENTED** [review]
Looks good. Just some tiny comments.

File: cpp/dolfinx/fem/dofmapbuilder.cpp (hunk 33 above)
Comment (chrisrichardson) [reply cpp/dolfinx/fem/dofmapbuilder.cpp:539]: Maybe. In another PR. I'd like to just get this merged.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**